feat/currency into main

kasou-sekai opened the pull request:
# Description

Implemented the currency system. This includes:

- Defining the core logic and data structure
- A CurrencyService to manage the global currency, ensuring consistent access and updates
- A Sun generator that periodically spawns Suns at randomised time intervals
- Interactive Sun entities that can be clicked to collect currency

Fixes / Closes # (issue)
Closes #23 
Closes #49 
Closes #52 
Closes #53 
Closes #55 
Closes #56 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

- [x] New feature (non-breaking change which adds functionality)


# How Has This Been Tested?

Automated unit tests cover the CurrencyService and the generator logic.
Functional testing was performed by running the game: Suns spawn at set intervals, fall with rotation, and stop rotating once clicked, correctly updating the currency value.

# Checklist:

- [x] My code follows the style guidelines of this project

- [x] I have performed a self-review of my code

- [x] I have commented my code, particularly in hard-to-understand areas

- [x] I have made corresponding changes to the documentation

- [x] My changes generate no new warnings

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] New and existing unit tests pass locally with my changes

- [x] Any dependent changes have been merged and published in downstream modules

pthneo reviewed (commented):
First of all, ran the game and it visually works well, and looks cool!

However, in terms of issues there's a memory leak in CurrencyGeneratorComponent where you create the forever action as it is never disposed of. From a quick chatgpt of the issue, you can add a reference to the action as a private variable and then remove the action later. 

Secondly it's probably good practice to add logger statements whenever something isn't going as expected ie when rs or stage is null you simply return but might be worth adding logger warn statements.

Lastly, please add javadoc comments in future, such that the javadocs website can be created.

BiancaRose2 reviewed (changes requested):
Missing various docstrings throughout - best to add for others' understanding
Missing dispose() methods in CurrencyGeneratorComponent, CurrencyInteraction, SunlightComponent
In CurrencyGeneratorComponent I don't think you deal with suns not being collected (ie clicked on) so it could keep generating and fill the screen with suns. Perhaps add an expiry, or otherwise a limit of spawning?
Missing loggers throughout - would be better to have for others' understanding and for debugging purposes when everyone's trying to merge
Also what is the purpose of keeping the PlayerStatsDisplay class? (Given all the content is deleted)

kasou-sekai commented:
> First of all, ran the game and it visually works well, and looks cool!
> 
> However, in terms of issues there's a memory leak in CurrencyGeneratorComponent where you create the forever action as it is never disposed of. From a quick chatgpt of the issue, you can add a reference to the action as a private variable and then remove the action later.
> 
> Secondly it's probably good practice to add logger statements whenever something isn't going as expected ie when rs or stage is null you simply return but might be worth adding logger warn statements.
> 
> Lastly, please add javadoc comments in future, such that the javadocs website can be created.



> Missing various docstrings throughout - best to add for others' understanding Missing dispose() methods in CurrencyGeneratorComponent, CurrencyInteraction, SunlightComponent In CurrencyGeneratorComponent I don't think you deal with suns not being collected (ie clicked on) so it could keep generating and fill the screen with suns. Perhaps add an expiry, or otherwise a limit of spawning? Missing loggers throughout - would be better to have for others' understanding and for debugging purposes when everyone's trying to merge Also what is the purpose of keeping the PlayerStatsDisplay class? (Given all the content is deleted)

thanks for the feedback. i've completed some requested changed:
 - fixed the memory leak in CurrencyGeneratorComponent by keeping a reference to the repeating action and removing it in dispose()
 - added auto-expiry for uncollected suns (they now fade out and are removed after a short TTL)
 - added some Javadoc comments
 - added logger warn statements

please take another look and let me know if anything else needs adjustment.
